### DISCUSSION TITLE
JSON layout support

### DISCUSSION THREAD

**Discussion by toncid** [okuramasafumi/alba#142] [Ideas]
Hello,

Great gem you have here! 👍 

I was wondering if there is a way to support simple "layouts" in a way that we would have a predefined JSON template with a payload placeholder within it?

The template could look like this:

```javascript
{
  status: {
    code: 200,
    message: "OK"
  },

  // payload goes here

  pagination: {
    nextPage: "..."
  }
}
```

The payload would then be yielded from a resource (properly inferred), like:

```javascript
{
  // status

  user: {...}

  // pagination
}
```

Or:

```javascript
{
  // status

  payments: [{...}]

  // pagination
}
```

Or even come with multiple subfields:

```javascript
{
  // status

  user: {...},
  payments: [{...}]

  // pagination
}
```

I haven't found a way to accomplish this with the current gem version (v1.4.0).

**Comment by okuramasafumi** [comment]
Hi, I've not considered JSON template with Alba, but it sounds interesting.
Do you come up with more concrete example of how to use template?

**Reply by toncid** [reply]
We have an API that we would like to migrate away from Jbuilder (sloooow), but also preserve as much of its flexibility as possible.

All our JSON responses share the same structure (similar to the template below), so the client can be sure that hidden fields are always supplied and the payload field name and data depend on the API endpoint (users, customers, merchants, payments, products, etc). This was relatively easy to accomplish with Jbuilder.

To further the proposal, the template could contain common fields from payload objects, so that actual model serializers only focus on their specific fields:

```javascript
{
  status: {...},

  "<object>": {
    id: "...",

    // other fields are supplied by the resource-specific serializer

    createdAt: "...",
    updatedAt: "..."
  },

  pagination: {...}
}
```

For collections, the output is like:

```javascript
{
  status: {...},

  "<objects>": [{
    id: "...",

    // other fields are supplied by the resource-specific serializer

    createdAt: "...",
    updatedAt: "..."
  }],

  pagination: {...}
}
```

**Reply by toncid** [reply]
In a way, this would be a request to support “partial” resources (similar to how Jbuilder partials extend the structure) in order to eliminate duplication and tighten the JSON structure control.

**Reply by okuramasafumi** [reply]
Does inheritance fulfill your needs?

```ruby
require 'alba'

class GenericResource
  include Alba::Resource

  attributes :id
  attribute :createdAt, &:created_at
  attribute :updatedAt, &:updated_at

  meta do
    {
      status: {
        code: 200,
        message: 'ok'
      },
      pagination: {
        page: params[:pagination][:page]
      }
    }
  end
end

class UserResource < GenericResource
  root_key :user, :users

  attributes :username
end

class User
  attr_reader :id, :username, :created_at, :updated_at

  def initialize(id, username)
    @id = id
    @username = username
    @created_at = Time.now
    @updated_at = Time.now
  end
end

user1 = User.new(1, 'User 1')
user2 = User.new(2, 'User 2')
users = [user1, user2]
p UserResource.new(users, params: {pagination: {page: 2}}).serialize
# => "{\"users\":[{\"id\":1,\"createdAt\":\"2021-07-01 17:20:41 +0900\",\"updatedAt\":\"2021-07-01 17:20:41 +0900\",\"username\":\"User 1\"},{\"id\":2,\"createdAt\":\"2021-07-01 17:20:41 +0900\",\"updatedAt\":\"2021-07-01 17:20:41 +0900\",\"username\":\"User 2\"}],\"meta\":{\"status\":{\"code\":200,\"message\":\"ok\"},\"pagination\":{\"page\":2}}}"
```

The output is not exactly what you expect, but this illustrates how Alba works with abstractions and metadata.

**Reply by toncid** [reply]
Thanks, this is definitely a move into the right direction, just not there yet. Changing the JSON structure is not a feasible option for us, hence the replacement should be completely invisible from the outside.

Based on my research, the proposed inheritance approach comes with the following limitations, i.e. I was not able to find a workaround:

* `meta` fields are added as a sub-object and there is no way to blend them into the main object.
* `meta` blocks override each other, i.e. there is no way to have them appear in their respective objects (`GenericResource` to modify the top object, `UserResource` to apply only to the `User` representation). Some way of setting arbitrary fields at the current level would be nice to have.
* It is unclear what needs to be done in the `meta` block for collections. The below code will yield `undefined method 'username' for #<Array:...> (NoMethodError)` when serializing a collection:

```ruby
class UserResource < GenericResource
  root_key :user, :users

  meta do
      {
        display_name: object.username,
      }
  end

  attributes :username
end
```

* For some reason, I look at `Resource` objects as how a single object is translated into JSON, regardless of it being part of a collection or a simple instance.
* Have you considered supporting proc or symbol shorthands for `if:` options?

```ruby
  attributes :date_of_birth, :gender, if: &:show_details? # or even: if: :show_details?
```

**Reply by okuramasafumi** [reply]
> * `meta` fields are added as a sub-object and there is no way to blend them into the main object.

Yes, that's the definition of "meta" in Alba. If you'd like to include some data in the main object, you can use methods such as `attributes`.

> * `meta` blocks override each other, i.e. there is no way to have them appear in their respective objects (`GenericResource` to modify the top object, `UserResource` to apply only to the `User` representation). Some way of setting arbitrary fields at the current level would be nice to have.

I understand this as a proposal to have "nested" resources. You can use `one` method to have nested resource.

> * It is unclear what needs to be done in the `meta` block for collections. The below code will yield `undefined method 'username' for #<Array:...> (NoMethodError)` when serializing a collection

In this example, `object` is an array of users. I guess `display_name` should be different for each user object, so having one `display_name` as metadata along with collection of users is not reasonable.

> * For some reason, I look at `Resource` objects as how a single object is translated into JSON, regardless of it being part of a collection or a simple instance.

By design, Alba uses the same resource for both singular resource and collection resource.

> * Have you considered supporting proc or symbol shorthands for `if:` options?

Currently Alba supports proc in `if` option. The code you gave causes `SyntaxError` so it's impossible to achieve. We can have symbol in `if` option but as you pointed out it's just a shorthand, so it's not a feature with high priority.

**Reply by toncid** [reply]
> We can have symbol in `if` option but as you pointed out it's just a shorthand, so it's not a feature with high priority.

Understood. My take is that reducing the amount boilerplate code is usually a good thing. I also agree that automatic symbol-to-proc expansion is not a big deal, but it would be nice to have. 

> I understand this as a proposal to have "nested" resources. You can use `one` method to have nested resource.

I didn’t quite get how `one` can be used to produce an arbitrary nested resource. Could you provide a simple example, please? Could it be somehow used to imitate the template from the beginning of this thread?

———

I am more concerned with the fact that Alba appears as a very opinionated JSON serializer (JSON:API is a no-no, but kind of imposes its own custom structure). General-purpose serializers should allow programmers to produce any format they wish without too much hassle. 

I have had really good time playing around with Alba (very fun to use - kudos!) while researching Jbuilder alternatives, but right now it feels far too restrictive. It is all legitimate, I am not criticizing, but one could expect more from “the fastest JSON serializer” out there.

**Reply by okuramasafumi** [reply]
> My take is that reducing the amount boilerplate code is usually a good thing

Yes, I think it's quite acceptable. It's just a matter of timing.

> Could it be somehow used to imitate the template from the beginning of this thread?

No, using `one` is not an alternative to templates.
Maybe I misunderstood something. I conclude that we need some sort of template to fulfill your needs.

> I am more concerned with the fact that Alba appears as a very opinionated JSON serializer

Right, Alba's core functionality relies on the concept of "serializing a resource into JSON". Alba doesn't support other JSON structures.
However, it doesn't mean that we shouldn't support it. We'd rather enable other types of JSON serialization so that users of jbuilder and other serializer can migrate to Alba easily.

**Comment by okuramasafumi** [comment]
Hi @toncid I've implemented layout feature in #160 , is it something you expected?

**Comment by toncid** [comment]
Hello @okuramasafumi! Wow, thanks! It definitely looks good and like a promising solution for layouts. How can it be tested?

**Reply by okuramasafumi** [reply]
https://github.com/okuramasafumi/sample_app_for_alba

This sample repository is where you can test it. Just clone it, run `bin/setup`, `rails s` and access `http://localhost:3000/users` and you'll see the response within a layout!
